Sony Xperia ZR2 Compact  - Pixel IMS stops

**Disclaimer**
Sorry to taking your attention to non-pixel phone.  However there are a number of threads on XDA that illustrates that your beatiful Pixel IMS works on other handsets too  - Sonyy Readme etc.  Hence I am asking for a help. Would be possible for you to help with troubleshooting why the Pixel IMS suddenly stops?

**Describe the bug**
Sony Xperia ZR2 Compact  - Pixel IMS stops.  


**To Reproduce**
Steps to reproduce the behavior:
1. Follow the instruction step by step up as published at  https://github.com/kyujin-cho/pixel-volte-patch/blob/main/README.en.md
up to "Tap "Allow all the time" when seeing prompt asking for Shizuku permission."  I.e. get shizuku running, Pixel IMS installed, all good, no issues.
3. Tap "Allow all the time" 
4. See error: the phone have a pop-up saying: 
    Pixel IMS keeps stopping
          - app info
          - close app

Any further attempt to restart the application ends up with the same error message as above.

**Expected behavior**
Pixel IMS continue to work or provides a description/log what went wrong and why it stops 

**Screenshots**
N/A 

**Desktop (please complete the following information):**
 - OS: Win 7 SP1
 - ADB platform-tools_r34.0.4-windows  (latest for Win 7)

**Smartphone (please complete the following information):**
 - Device: Sony Xperia XZ2 Compact , H8324
 - OS: Android 10 
 - Firmware: H8324_Customized AU_52.1.A.3.49-R4C

**Additional context**
Mobile Network : Internode 
APN Vodafone live! live.vodafone.com
Phone / SIM are not locked

Here it's mine with exactly the same error:

Device: Huawei P30 Pro
OS: Android 10


The logcat says:

`03-12 00:50:13.060  5086  5086 E AndroidRuntime: FATAL EXCEPTION: main
03-12 00:50:13.060  5086  5086 E AndroidRuntime: Process: dev.bluehouse.enablevolte, PID: 5086
03-12 00:50:13.060  5086  5086 E AndroidRuntime: java.lang.NoClassDefFoundError: Failed resolution of: Landroid/telephony/TelephonyFrameworkInitializer;
03-12 00:50:13.060  5086  5086 E AndroidRuntime:        at dev.bluehouse.enablevolte.Moder.getSub(Moder.kt:72)
03-12 00:50:13.060  5086  5086 E AndroidRuntime:        at dev.bluehouse.enablevolte.CarrierModer.getSubscriptions(Moder.kt:87)
03-12 00:50:13.060  5086  5086 E AndroidRuntime:        at dev.bluehouse.enablevolte.HomeActivityKt.PixelIMSApp$loadApplication(HomeActivity.kt:133)
03-12 00:50:13.060  5086  5086 E AndroidRuntime:        at dev.bluehouse.enablevolte.HomeActivityKt.access$PixelIMSApp$loadApplication(HomeActivity.kt:1)
03-12 00:50:13.060  5086  5086 E AndroidRuntime:        at dev.bluehouse.enablevolte.HomeActivityKt$PixelIMSApp$1.invoke(HomeActivity.kt:156)
03-12 00:50:13.060  5086  5086 E AndroidRuntime:        at dev.bluehouse.enablevolte.HomeActivityKt$PixelIMSApp$1.invoke(HomeActivity.kt:154)
03-12 00:50:13.060  5086  5086 E AndroidRuntime:        at dev.bluehouse.enablevolte.ComponentsKt$OnLifecycleEvent$1$1$observer$1.onStateChanged(Components.kt:69)
03-12 00:50:13.060  5086  5086 E AndroidRuntime:        at androidx.lifecycle.LifecycleRegistry$ObserverWithState.dispatchEvent(LifecycleRegistry.kt:314)
03-12 00:50:13.060  5086  5086 E AndroidRuntime:        at androidx.lifecycle.LifecycleRegistry.forwardPass(LifecycleRegistry.kt:251)
03-12 00:50:13.060  5086  5086 E AndroidRuntime:        at androidx.lifecycle.LifecycleRegistry.sync(LifecycleRegistry.kt:287)
03-12 00:50:13.060  5086  5086 E AndroidRuntime:        at androidx.lifecycle.LifecycleRegistry.addObserver(LifecycleRegistry.kt:199)
03-12 00:50:13.060  5086  5086 E AndroidRuntime:        at androidx.compose.ui.platform.WrappedComposition$setContent$1.invoke(Wrapper.android.kt:138)
03-12 00:50:13.060  5086  5086 E AndroidRuntime:        at androidx.compose.ui.platform.WrappedComposition$setContent$1.invoke(Wrapper.android.kt:131)
03-12 00:50:13.060  5086  5086 E AndroidRuntime:        at androidx.compose.ui.platform.AndroidComposeView.onAttachedToWindow(AndroidComposeView.android.kt:1147)
03-12 00:50:13.060  5086  5086 E AndroidRuntime:        at android.view.View.dispatchAttachedToWindow(View.java:20107)
03-12 00:50:13.060  5086  5086 E AndroidRuntime:        at android.view.ViewGroup.dispatchAttachedToWindow(ViewGroup.java:3619)
03-12 00:50:13.060  5086  5086 E AndroidRuntime:        at android.view.ViewGroup.dispatchAttachedToWindow(ViewGroup.java:3626)
03-12 00:50:13.060  5086  5086 E AndroidRuntime:        at android.view.ViewGroup.dispatchAttachedToWindow(ViewGroup.java:3626)
03-12 00:50:13.060  5086  5086 E AndroidRuntime:        at android.view.ViewGroup.dispatchAttachedToWindow(ViewGroup.java:3626)
03-12 00:50:13.060  5086  5086 E AndroidRuntime:        at android.view.ViewGroup.dispatchAttachedToWindow(ViewGroup.java:3626)
03-12 00:50:13.060  5086  5086 E AndroidRuntime:        at android.view.ViewRootImpl.performTraversals(ViewRootImpl.java:2549)
03-12 00:50:13.060  5086  5086 E AndroidRuntime:        at android.view.ViewRootImpl.doTraversal(ViewRootImpl.java:2166)
03-12 00:50:13.060  5086  5086 E AndroidRuntime:        at android.view.ViewRootImpl$TraversalRunnable.run(ViewRootImpl.java:8887)
03-12 00:50:13.060  5086  5086 E AndroidRuntime:        at android.view.Choreographer$CallbackRecord.run(Choreographer.java:1280)
03-12 00:50:13.060  5086  5086 E AndroidRuntime:        at android.view.Choreographer.doCallbacks(Choreographer.java:1019)
03-12 00:50:13.060  5086  5086 E AndroidRuntime:        at android.view.Choreographer.doFrame(Choreographer.java:911)
03-12 00:50:13.060  5086  5086 E AndroidRuntime:        at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:1248)
03-12 00:50:13.060  5086  5086 E AndroidRuntime:        at android.os.Handler.handleCallback(Handler.java:900)
03-12 00:50:13.060  5086  5086 E AndroidRuntime:        at android.os.Handler.dispatchMessage(Handler.java:103)
03-12 00:50:13.060  5086  5086 E AndroidRuntime:        at android.os.Looper.loop(Looper.java:219)
03-12 00:50:13.060  5086  5086 E AndroidRuntime:        at android.app.ActivityThread.main(ActivityThread.java:8668)
03-12 00:50:13.060  5086  5086 E AndroidRuntime:        at java.lang.reflect.Method.invoke(Native Method)
03-12 00:50:13.060  5086  5086 E AndroidRuntime:        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:513)
03-12 00:50:13.060  5086  5086 E AndroidRuntime:        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1109)
03-12 00:50:13.060  5086  5086 E AndroidRuntime: Caused by: java.lang.ClassNotFoundException: android.telephony.TelephonyFrameworkInitializer
03-12 00:50:13.060  5086  5086 E AndroidRuntime:        ... 34 more
`

The problems seems related with the library used in Moder.kt: [import android.telephony.TelephonyFrameworkInitializer](https://github.com/kyujin-cho/pixel-volte-patch/blob/a944bd127c88066403a9a5e3c897bac989cb8eb7/app/src/main/java/dev/bluehouse/enablevolte/Moder.kt#L11C1-L11C55) maybe as @bolabola6789 mentioned, it's not compatible with Android 10, or maybe the sdk used is too recent for us [API level 34 I think (Android 14).](https://developer.android.com/guide/topics/manifest/uses-sdk-element?hl=es-419#ApiLevels) I'm not familiar with programming at this level, just starting in this world, but who knows, this comment might help to another in something.

> Here it's mine with exactly the same error:
> 
> Device: Huawei P30 Pro OS: Android 10
> 
> The logcat says:
> 
> ```
> 03-12 00:50:13.060  5086  5086 E AndroidRuntime: FATAL EXCEPTION: main
> 03-12 00:50:13.060  5086  5086 E AndroidRuntime: Process: dev.bluehouse.enablevolte, PID: 5086
> 03-12 00:50:13.060  5086  5086 E AndroidRuntime: java.lang.NoClassDefFoundError: Failed resolution of: Landroid/telephony/TelephonyFrameworkInitializer;
> 03-12 00:50:13.060  5086  5086 E AndroidRuntime:        at dev.bluehouse.enablevolte.Moder.getSub(Moder.kt:72)
> 03-12 00:50:13.060  5086  5086 E AndroidRuntime:        at dev.bluehouse.enablevolte.CarrierModer.getSubscriptions(Moder.kt:87)
> 03-12 00:50:13.060  5086  5086 E AndroidRuntime:        at dev.bluehouse.enablevolte.HomeActivityKt.PixelIMSApp$loadApplication(HomeActivity.kt:133)
> 03-12 00:50:13.060  5086  5086 E AndroidRuntime:        at dev.bluehouse.enablevolte.HomeActivityKt.access$PixelIMSApp$loadApplication(HomeActivity.kt:1)
> 03-12 00:50:13.060  5086  5086 E AndroidRuntime:        at dev.bluehouse.enablevolte.HomeActivityKt$PixelIMSApp$1.invoke(HomeActivity.kt:156)
> 03-12 00:50:13.060  5086  5086 E AndroidRuntime:        at dev.bluehouse.enablevolte.HomeActivityKt$PixelIMSApp$1.invoke(HomeActivity.kt:154)
> 03-12 00:50:13.060  5086  5086 E AndroidRuntime:        at dev.bluehouse.enablevolte.ComponentsKt$OnLifecycleEvent$1$1$observer$1.onStateChanged(Components.kt:69)
> 03-12 00:50:13.060  5086  5086 E AndroidRuntime:        at androidx.lifecycle.LifecycleRegistry$ObserverWithState.dispatchEvent(LifecycleRegistry.kt:314)
> 03-12 00:50:13.060  5086  5086 E AndroidRuntime:        at androidx.lifecycle.LifecycleRegistry.forwardPass(LifecycleRegistry.kt:251)
> 03-12 00:50:13.060  5086  5086 E AndroidRuntime:        at androidx.lifecycle.LifecycleRegistry.sync(LifecycleRegistry.kt:287)
> 03-12 00:50:13.060  5086  5086 E AndroidRuntime:        at androidx.lifecycle.LifecycleRegistry.addObserver(LifecycleRegistry.kt:199)
> 03-12 00:50:13.060  5086  5086 E AndroidRuntime:        at androidx.compose.ui.platform.WrappedComposition$setContent$1.invoke(Wrapper.android.kt:138)
> 03-12 00:50:13.060  5086  5086 E AndroidRuntime:        at androidx.compose.ui.platform.WrappedComposition$setContent$1.invoke(Wrapper.android.kt:131)
> 03-12 00:50:13.060  5086  5086 E AndroidRuntime:        at androidx.compose.ui.platform.AndroidComposeView.onAttachedToWindow(AndroidComposeView.android.kt:1147)
> 03-12 00:50:13.060  5086  5086 E AndroidRuntime:        at android.view.View.dispatchAttachedToWindow(View.java:20107)
> 03-12 00:50:13.060  5086  5086 E AndroidRuntime:        at android.view.ViewGroup.dispatchAttachedToWindow(ViewGroup.java:3619)
> 03-12 00:50:13.060  5086  5086 E AndroidRuntime:        at android.view.ViewGroup.dispatchAttachedToWindow(ViewGroup.java:3626)
> 03-12 00:50:13.060  5086  5086 E AndroidRuntime:        at android.view.ViewGroup.dispatchAttachedToWindow(ViewGroup.java:3626)
> 03-12 00:50:13.060  5086  5086 E AndroidRuntime:        at android.view.ViewGroup.dispatchAttachedToWindow(ViewGroup.java:3626)
> 03-12 00:50:13.060  5086  5086 E AndroidRuntime:        at android.view.ViewGroup.dispatchAttachedToWindow(ViewGroup.java:3626)
> 03-12 00:50:13.060  5086  5086 E AndroidRuntime:        at android.view.ViewRootImpl.performTraversals(ViewRootImpl.java:2549)
> 03-12 00:50:13.060  5086  5086 E AndroidRuntime:        at android.view.ViewRootImpl.doTraversal(ViewRootImpl.java:2166)
> 03-12 00:50:13.060  5086  5086 E AndroidRuntime:        at android.view.ViewRootImpl$TraversalRunnable.run(ViewRootImpl.java:8887)
> 03-12 00:50:13.060  5086  5086 E AndroidRuntime:        at android.view.Choreographer$CallbackRecord.run(Choreographer.java:1280)
> 03-12 00:50:13.060  5086  5086 E AndroidRuntime:        at android.view.Choreographer.doCallbacks(Choreographer.java:1019)
> 03-12 00:50:13.060  5086  5086 E AndroidRuntime:        at android.view.Choreographer.doFrame(Choreographer.java:911)
> 03-12 00:50:13.060  5086  5086 E AndroidRuntime:        at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:1248)
> 03-12 00:50:13.060  5086  5086 E AndroidRuntime:        at android.os.Handler.handleCallback(Handler.java:900)
> 03-12 00:50:13.060  5086  5086 E AndroidRuntime:        at android.os.Handler.dispatchMessage(Handler.java:103)
> 03-12 00:50:13.060  5086  5086 E AndroidRuntime:        at android.os.Looper.loop(Looper.java:219)
> 03-12 00:50:13.060  5086  5086 E AndroidRuntime:        at android.app.ActivityThread.main(ActivityThread.java:8668)
> 03-12 00:50:13.060  5086  5086 E AndroidRuntime:        at java.lang.reflect.Method.invoke(Native Method)
> 03-12 00:50:13.060  5086  5086 E AndroidRuntime:        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:513)
> 03-12 00:50:13.060  5086  5086 E AndroidRuntime:        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1109)
> 03-12 00:50:13.060  5086  5086 E AndroidRuntime: Caused by: java.lang.ClassNotFoundException: android.telephony.TelephonyFrameworkInitializer
> 03-12 00:50:13.060  5086  5086 E AndroidRuntime:        ... 34 more
> ```
> 
> The problem seems related to the library used in Moder.kt: [import android.telephony.TelephonyFrameworkInitializer](https://github.com/kyujin-cho/pixel-volte-patch/blob/a944bd127c88066403a9a5e3c897bac989cb8eb7/app/src/main/java/dev/bluehouse/enablevolte/Moder.kt#L11C1-L11C55) maybe as @bolabola6789 mentioned, it's not compatible with Android 10 (but the minimum SDK in Google play is 29 which is just for Android 10) I'm not familiar with programming at this level, just starting in this world, but who knows, this comment might help to another in something.

You might be onto something

--- a/app/src/main/java/dev/bluehouse/enablevolte/Moder.kt
+++ b/app/src/main/java/dev/bluehouse/enablevolte/Moder.kt
@@ -128,7 +128,8 @@ class CarrierModer(
             val sub = this.loadCachedInterface { sub }
             try {
                 return sub.getActiveSubscriptionInfoList(null, null, true) ?: emptyList()
-            } catch (e: NoSuchMethodError) {}
+            } catch (e: NoSuchMethodError) {
+            }
             return try {
                 val getActiveSubscriptionInfoListMethod =
                     sub.javaClass.getMethod(
@@ -398,10 +399,14 @@ class SubscriptionModer(
         return config.get(key)
     }
 
-    fun getConfigForSubId(iCclInstance: ICarrierConfigLoader, subscriptionId: Int): PersistableBundle {
+    fun getConfigForSubId(
+        iCclInstance: ICarrierConfigLoader,
+        subscriptionId: Int,
+    ): PersistableBundle {
         try {
             return iCclInstance.getConfigForSubIdWithFeature(subscriptionId, iCclInstance.defaultCarrierServicePackageName, "")
-        } catch (e: NoSuchMethodError) {}
+        } catch (e: NoSuchMethodError) {
+        }
         return try {
             iCclInstance.getConfigForSubId(subscriptionId, iCclInstance.defaultCarrierServicePackageName)
         } catch (e: NoSuchMethodError) {

--- a/app/src/main/java/dev/bluehouse/enablevolte/Moder.kt
+++ b/app/src/main/java/dev/bluehouse/enablevolte/Moder.kt
@@ -99,7 +99,7 @@ class CarrierModer(
         get() {
             val sub = this.loadCachedInterface { sub }
             try {
-                return sub.getActiveSubscriptionInfoList(null, null, true)
+                return sub.getActiveSubscriptionInfoList(null, null, true) ?: emptyList()
             } catch (e: NoSuchMethodError) {}
             return try {
                 val getActiveSubscriptionInfoListMethod =
@@ -108,14 +108,14 @@ class CarrierModer(
                         String::class.java,
                         String::class.java,
                     )
-                (getActiveSubscriptionInfoListMethod.invoke(sub, null, null) as List<SubscriptionInfo>)
+                (getActiveSubscriptionInfoListMethod.invoke(sub, null, null) as? List<SubscriptionInfo>) ?: emptyList()
             } catch (e: NoSuchMethodException) {
                 val getActiveSubscriptionInfoListMethod =
                     sub.javaClass.getMethod(
                         "getActiveSubscriptionInfoList",
                         String::class.java,
                     )
-                (getActiveSubscriptionInfoListMethod.invoke(sub, null) as List<SubscriptionInfo>)
+                (getActiveSubscriptionInfoListMethod.invoke(sub, null) as? List<SubscriptionInfo>) ?: emptyList()
             }
         }
 

--- a/app/src/main/java/dev/bluehouse/enablevolte/Moder.kt
+++ b/app/src/main/java/dev/bluehouse/enablevolte/Moder.kt
@@ -98,18 +98,24 @@ class CarrierModer(
     val subscriptions: List<SubscriptionInfo>
         get() {
             val sub = this.loadCachedInterface { sub }
+            try {
+                return sub.getActiveSubscriptionInfoList(null, null, true)
+            } catch (e: NoSuchMethodError) {}
             return try {
-                sub.getActiveSubscriptionInfoList(null, null, true)
-            } catch (e: NoSuchMethodError) {
-                // FIXME: lift up reflect as soon as official source code releases
                 val getActiveSubscriptionInfoListMethod =
                     sub.javaClass.getMethod(
                         "getActiveSubscriptionInfoList",
                         String::class.java,
                         String::class.java,
-                        Boolean::class.java,
                     )
-                (getActiveSubscriptionInfoListMethod.invoke(sub, null, null, false) as List<SubscriptionInfo>)
+                (getActiveSubscriptionInfoListMethod.invoke(sub, null, null) as List<SubscriptionInfo>)
+            } catch (e: NoSuchMethodException) {
+                val getActiveSubscriptionInfoListMethod =
+                    sub.javaClass.getMethod(
+                        "getActiveSubscriptionInfoList",
+                        String::class.java,
+                    )
+                (getActiveSubscriptionInfoListMethod.invoke(sub, null) as List<SubscriptionInfo>)
             }
         }
 

--- a/app/src/main/java/dev/bluehouse/enablevolte/Moder.kt
+++ b/app/src/main/java/dev/bluehouse/enablevolte/Moder.kt
@@ -460,7 +460,7 @@ class SubscriptionModer(
     val wfcSpnFormatIndex: Int
         get() = this.getIntValue(CarrierConfigManager.KEY_WFC_SPN_FORMAT_IDX_INT)
 
-    val carrierName: String
+    val carrierName: String?
         get() = this.loadCachedInterface { telephony }.getSubscriptionCarrierName(this.subscriptionId)
 
     val showVoWifiIcon: Boolean

--- a/app/src/main/java/dev/bluehouse/enablevolte/Moder.kt
+++ b/app/src/main/java/dev/bluehouse/enablevolte/Moder.kt
@@ -9,6 +9,7 @@ import android.os.Build
 import android.os.Build.VERSION_CODES
 import android.os.Bundle
 import android.os.IInterface
+import android.os.PersistableBundle
 import android.os.ServiceManager
 import android.telephony.CarrierConfigManager
 import android.telephony.SubscriptionInfo
@@ -309,7 +310,7 @@ class SubscriptionModer(
         }
         val iCclInstance = this.loadCachedInterface { carrierConfigLoader }
 
-        val config = iCclInstance.getConfigForSubIdWithFeature(subscriptionId, iCclInstance.defaultCarrierServicePackageName, "")
+        val config = this.getConfigForSubId(iCclInstance, subscriptionId)
         return config.getString(key)
     }
 
@@ -321,7 +322,7 @@ class SubscriptionModer(
         }
         val iCclInstance = this.loadCachedInterface { carrierConfigLoader }
 
-        val config = iCclInstance.getConfigForSubIdWithFeature(subscriptionId, iCclInstance.defaultCarrierServicePackageName, "")
+        val config = this.getConfigForSubId(iCclInstance, subscriptionId)
         return config.getBoolean(key)
     }
 
@@ -333,7 +334,7 @@ class SubscriptionModer(
         }
         val iCclInstance = this.loadCachedInterface { carrierConfigLoader }
 
-        val config = iCclInstance.getConfigForSubIdWithFeature(subscriptionId, iCclInstance.defaultCarrierServicePackageName, "")
+        val config = this.getConfigForSubId(iCclInstance, subscriptionId)
         return config.getInt(key)
     }
 
@@ -345,7 +346,7 @@ class SubscriptionModer(
         }
         val iCclInstance = this.loadCachedInterface { carrierConfigLoader }
 
-        val config = iCclInstance.getConfigForSubIdWithFeature(subscriptionId, iCclInstance.defaultCarrierServicePackageName, "")
+        val config = this.getConfigForSubId(iCclInstance, subscriptionId)
         return config.getLong(key)
     }
 
@@ -357,7 +358,7 @@ class SubscriptionModer(
         }
         val iCclInstance = this.loadCachedInterface { carrierConfigLoader }
 
-        val config = iCclInstance.getConfigForSubIdWithFeature(subscriptionId, iCclInstance.defaultCarrierServicePackageName, "")
+        val config = this.getConfigForSubId(iCclInstance, subscriptionId)
         return config.getBooleanArray(key) ?: BooleanArray(0)
     }
 
@@ -369,7 +370,7 @@ class SubscriptionModer(
         }
         val iCclInstance = this.loadCachedInterface { carrierConfigLoader }
 
-        val config = iCclInstance.getConfigForSubIdWithFeature(subscriptionId, iCclInstance.defaultCarrierServicePackageName, "")
+        val config = this.getConfigForSubId(iCclInstance, subscriptionId)
         return config.getIntArray(key) ?: IntArray(0)
     }
 
@@ -381,7 +382,7 @@ class SubscriptionModer(
         }
         val iCclInstance = this.loadCachedInterface { carrierConfigLoader }
 
-        val config = iCclInstance.getConfigForSubIdWithFeature(subscriptionId, iCclInstance.defaultCarrierServicePackageName, "")
+        val config = this.getConfigForSubId(iCclInstance, subscriptionId)
         return config.getStringArray(key) ?: emptyArray()
     }
 
@@ -393,8 +394,24 @@ class SubscriptionModer(
         }
         val iCclInstance = this.loadCachedInterface { carrierConfigLoader }
 
-        val config = iCclInstance.getConfigForSubIdWithFeature(subscriptionId, iCclInstance.defaultCarrierServicePackageName, "")
+        val config = this.getConfigForSubId(iCclInstance, subscriptionId)
         return config.get(key)
+    }
+
+    fun getConfigForSubId(iCclInstance: ICarrierConfigLoader, subscriptionId: Int): PersistableBundle {
+        try {
+            return iCclInstance.getConfigForSubIdWithFeature(subscriptionId, iCclInstance.defaultCarrierServicePackageName, "")
+        } catch (e: NoSuchMethodError) {}
+        return try {
+            iCclInstance.getConfigForSubId(subscriptionId, iCclInstance.defaultCarrierServicePackageName)
+        } catch (e: NoSuchMethodError) {
+            val getConfigForSubIdMethod =
+                iCclInstance.javaClass.getMethod(
+                    "getConfigForSubId",
+                    Int::class.javaPrimitiveType,
+                )
+            (getConfigForSubIdMethod.invoke(iCclInstance, subscriptionId) as PersistableBundle)
+        }
     }
 
     val simSlotIndex: Int

--- a/app/src/main/java/dev/bluehouse/enablevolte/Moder.kt
+++ b/app/src/main/java/dev/bluehouse/enablevolte/Moder.kt
@@ -109,7 +109,17 @@ class CarrierModer(
 ) : Moder() {
     fun getActiveSubscriptionInfoForSimSlotIndex(index: Int): SubscriptionInfo? {
         val sub = this.loadCachedInterface { sub }
-        return sub.getActiveSubscriptionInfoForSimSlotIndex(index, null, null)
+        return try {
+            sub.getActiveSubscriptionInfoForSimSlotIndex(index, null, null)
+        } catch (e: NoSuchMethodError) {
+            val getActiveSubscriptionInfoForSimSlotIndexMethod =
+                sub.javaClass.getMethod(
+                    "getActiveSubscriptionInfoForSimSlotIndex",
+                    Int::class.javaPrimitiveType,
+                    String::class.java,
+                )
+            (getActiveSubscriptionInfoForSimSlotIndexMethod.invoke(sub, index, null) as? SubscriptionInfo)
+        }
     }
 
     val subscriptions: List<SubscriptionInfo>

--- a/app/src/main/java/dev/bluehouse/enablevolte/Moder.kt
+++ b/app/src/main/java/dev/bluehouse/enablevolte/Moder.kt
@@ -312,7 +312,7 @@ class SubscriptionModer(
         val iCclInstance = this.loadCachedInterface { carrierConfigLoader }
 
         val config = this.getConfigForSubId(iCclInstance, subscriptionId)
-        return config.getString(key)
+        return config?.getString(key)
     }
 
     fun getBooleanValue(key: String): Boolean {
@@ -324,7 +324,7 @@ class SubscriptionModer(
         val iCclInstance = this.loadCachedInterface { carrierConfigLoader }
 
         val config = this.getConfigForSubId(iCclInstance, subscriptionId)
-        return config.getBoolean(key)
+        return config?.getBoolean(key) ?: false
     }
 
     fun getIntValue(key: String): Int {
@@ -336,7 +336,7 @@ class SubscriptionModer(
         val iCclInstance = this.loadCachedInterface { carrierConfigLoader }
 
         val config = this.getConfigForSubId(iCclInstance, subscriptionId)
-        return config.getInt(key)
+        return config?.getInt(key) ?: -1
     }
 
     fun getLongValue(key: String): Long {
@@ -348,7 +348,7 @@ class SubscriptionModer(
         val iCclInstance = this.loadCachedInterface { carrierConfigLoader }
 
         val config = this.getConfigForSubId(iCclInstance, subscriptionId)
-        return config.getLong(key)
+        return config?.getLong(key) ?: -1L
     }
 
     fun getBooleanArrayValue(key: String): BooleanArray {
@@ -360,7 +360,7 @@ class SubscriptionModer(
         val iCclInstance = this.loadCachedInterface { carrierConfigLoader }
 
         val config = this.getConfigForSubId(iCclInstance, subscriptionId)
-        return config.getBooleanArray(key) ?: BooleanArray(0)
+        return config?.getBooleanArray(key) ?: BooleanArray(0)
     }
 
     fun getIntArrayValue(key: String): IntArray {
@@ -372,7 +372,7 @@ class SubscriptionModer(
         val iCclInstance = this.loadCachedInterface { carrierConfigLoader }
 
         val config = this.getConfigForSubId(iCclInstance, subscriptionId)
-        return config.getIntArray(key) ?: IntArray(0)
+        return config?.getIntArray(key) ?: IntArray(0)
     }
 
     fun getStringArrayValue(key: String): Array<String> {
@@ -384,7 +384,7 @@ class SubscriptionModer(
         val iCclInstance = this.loadCachedInterface { carrierConfigLoader }
 
         val config = this.getConfigForSubId(iCclInstance, subscriptionId)
-        return config.getStringArray(key) ?: emptyArray()
+        return config?.getStringArray(key) ?: emptyArray()
     }
 
     fun getValue(key: String): Any? {
@@ -396,13 +396,13 @@ class SubscriptionModer(
         val iCclInstance = this.loadCachedInterface { carrierConfigLoader }
 
         val config = this.getConfigForSubId(iCclInstance, subscriptionId)
-        return config.get(key)
+        return config?.get(key)
     }
 
     fun getConfigForSubId(
         iCclInstance: ICarrierConfigLoader,
         subscriptionId: Int,
-    ): PersistableBundle {
+    ): PersistableBundle? {
         try {
             return iCclInstance.getConfigForSubIdWithFeature(subscriptionId, iCclInstance.defaultCarrierServicePackageName, "")
         } catch (e: NoSuchMethodError) {
@@ -415,7 +415,7 @@ class SubscriptionModer(
                     "getConfigForSubId",
                     Int::class.javaPrimitiveType,
                 )
-            (getConfigForSubIdMethod.invoke(iCclInstance, subscriptionId) as PersistableBundle)
+            (getConfigForSubIdMethod.invoke(iCclInstance, subscriptionId) as? PersistableBundle)
         }
     }
 

--- a/app/src/main/java/dev/bluehouse/enablevolte/Moder.kt
+++ b/app/src/main/java/dev/bluehouse/enablevolte/Moder.kt
@@ -9,6 +9,7 @@ import android.os.Build
 import android.os.Build.VERSION_CODES
 import android.os.Bundle
 import android.os.IInterface
+import android.os.ServiceManager
 import android.telephony.CarrierConfigManager
 import android.telephony.SubscriptionInfo
 import android.telephony.TelephonyFrameworkInitializer
@@ -46,10 +47,14 @@ open class Moder {
         get() =
             ICarrierConfigLoader.Stub.asInterface(
                 ShizukuBinderWrapper(
-                    TelephonyFrameworkInitializer
-                        .getTelephonyServiceManager()
-                        .carrierConfigServiceRegisterer
-                        .get()!!,
+                    try {
+                        TelephonyFrameworkInitializer
+                            .getTelephonyServiceManager()
+                            .carrierConfigServiceRegisterer
+                            .get()
+                    } catch (e: NoClassDefFoundError) {
+                        ServiceManager.getService(Context.CARRIER_CONFIG_SERVICE)
+                    }!!,
                 ),
             )
 
@@ -57,10 +62,14 @@ open class Moder {
         get() =
             ITelephony.Stub.asInterface(
                 ShizukuBinderWrapper(
-                    TelephonyFrameworkInitializer
-                        .getTelephonyServiceManager()
-                        .telephonyServiceRegisterer
-                        .get()!!,
+                    try {
+                        TelephonyFrameworkInitializer
+                            .getTelephonyServiceManager()
+                            .telephonyServiceRegisterer
+                            .get()
+                    } catch (e: NoClassDefFoundError) {
+                        ServiceManager.getService(Context.TELEPHONY_SERVICE)
+                    }!!,
                 ),
             )
 
@@ -68,10 +77,14 @@ open class Moder {
         get() =
             IPhoneSubInfo.Stub.asInterface(
                 ShizukuBinderWrapper(
-                    TelephonyFrameworkInitializer
-                        .getTelephonyServiceManager()
-                        .phoneSubServiceRegisterer
-                        .get()!!,
+                    try {
+                        TelephonyFrameworkInitializer
+                            .getTelephonyServiceManager()
+                            .phoneSubServiceRegisterer
+                            .get()
+                    } catch (e: NoClassDefFoundError) {
+                        ServiceManager.getService("iphonesubinfo")
+                    }!!,
                 ),
             )
 
@@ -79,10 +92,14 @@ open class Moder {
         get() =
             ISub.Stub.asInterface(
                 ShizukuBinderWrapper(
-                    TelephonyFrameworkInitializer
-                        .getTelephonyServiceManager()
-                        .subscriptionServiceRegisterer
-                        .get()!!,
+                    try {
+                        TelephonyFrameworkInitializer
+                            .getTelephonyServiceManager()
+                            .subscriptionServiceRegisterer
+                            .get()
+                    } catch (e: NoClassDefFoundError) {
+                        ServiceManager.getService("isub")
+                    }!!,
                 ),
             )
 }

